File: Python/1_Creational Patterns/4_Builder/Pizza_Order.py
class Pizza:
    class PizzaBuilder:
        def __init__(self, size):
            self.size = size
            self.crust = ""
            self.cheese = ""
            self.toppings = []

        def crust(self, crust):
            self.crust = crust
            return self

        def cheese(self, cheese):
            self.cheese = cheese
            return self

        def add_topping(self, topping):
            self.toppings.append(topping)
            return self

        def build(self):
            return Pizza(self)

    def __init__(self, builder):
        self.size = builder.size
        self.crust = builder.crust
        self.cheese = builder.cheese
        self.toppings = builder.toppings

    def __str__(self):
        return f"Pizza(size='{self.size}', crust='{self.crust}', cheese='{self.cheese}', toppings={self.toppings})"

if __name__ == "__main__":
    pizza = Pizza.PizzaBuilder("Large") \
        .crust("Thin") \
        .cheese("Mozzarella") \
        .add_topping("Pepperoni") \
        .add_topping("Mushrooms") \
        .add_topping("Onions") \
        .build()

    print(pizza)

File: Java/1_Creational Patterns/4_Builder/Pizza_Order.java
// Pizza.java

import java.util.List;

public class Pizza {
    private final String size;
    private final String crust;
    private final String cheese;
    private final List<String> toppings;

    private Pizza(PizzaBuilder builder) {
        this.size = builder.size;
        this.crust = builder.crust;
        this.cheese = builder.cheese;
        this.toppings = builder.toppings;
    }

    public static class PizzaBuilder {
        private final String size;
        private String crust;
        private String cheese;
        private List<String> toppings = new ArrayList<>();

        public PizzaBuilder(String size) {
            this.size = size;
        }

        public PizzaBuilder crust(String crust) {
            this.crust = crust;
            return this;
        }

        public PizzaBuilder cheese(String cheese) {
            this.cheese = cheese;
            return this;
        }

        public PizzaBuilder addTopping(String topping) {
            this.toppings.add(topping);
            return this;
        }

        public Pizza build() {
            return new Pizza(this);
        }
    }

    @Override
    public String toString() {
        return "Pizza{" +
                "size='" + size + '\'' +
                ", crust='" + crust + '\'' +
                ", cheese='" + cheese + '\'' +
                ", toppings=" + toppings +
                '}';
    }
}

// Main.java
public class Main {
    public static void main(String[] args) {
        Pizza pizza = new Pizza.PizzaBuilder("Large")
                .crust("Thin")
                .cheese("Mozzarella")
                .addTopping("Pepperoni")
                .addTopping("Mushrooms")
                .addTopping("Onions")
                .build();

        System.out.println(pizza);
    }
}
